### ISSUE TITLE
`docker commit --change` doesn't work for `ADD`, `COPY`, `RUN` despite docs' claim

### ISSUE THREAD

**Issue by posita** [docker/docker#13886]
From <https://docs.docker.com/reference/commandline/cli/#commit>:

> The `--change` option will apply `Dockerfile` instructions to the image that is created. Supported `Dockerfile` instructions: `ADD`|...|`RUN`|...|`COPY`

But:

```sh
% docker version
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 7c8fca2
OS/Arch (client): darwin/amd64
Server version: 1.6.0
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 4749651
OS/Arch (server): linux/amd64
% docker create debian:jessie
6643...3c29
% echo foo >foo
% docker commit -c 'ADD foo /tmp/' 6643...3c29
FATA[0000] Error response from daemon: add is not a valid change command
% docker commit -c 'RUN ls /tmp/' 6643...3c29
FATA[0000] Error response from daemon: run is not a valid change command
% docker commit -c 'COPY foo /tmp/' 6643...3c29
FATA[0000] Error response from daemon: copy is not a valid change command
```

Caveat: there might be others.

**Comment by duglin** [comment]
I think this is a dup of: https://github.com/docker/docker/pull/13744  if not please reopen.

**Comment by posita** [comment]
I saw that issue originally, but interpreted it as limited to the `LABEL` command. Your https://github.com/docker/docker/pull/13744#issuecomment-111193218 requiring inclusion of the commands covered here seems to make it a dup (if not in name, than in substance). :+1:

**Comment by cpuguy83** [comment]
Reopining because this is not really a dup.

**Comment by jfrazelle** [comment]
closed by https://github.com/docker/docker/pull/14501

**Comment by posita** [comment]
@jfrazelle, I might have misunderstood, but how does #14301 resolve this issue? Is it being resolved because `ADD`, `COPY`, and `RUN` have been yanked from the docs (rather than properly supporting them from the `commit` command)? If so, should I open another issue to request that support for those be added?

**Comment by cpuguy83** [comment]
@posita `RUN` doesn't really make sense on commit.
`ADD` and `COPY` are also a bit weird, because presumably these files should be there _before_ the container is started.

There's also #13171, which enables copying into a container, which also opens up the door for a full client side builder.

**Comment by posita** [comment]
@cpuguy83, thanks! I was unaware of the proposed `cp` command.

I think `RUN` would make sense if you're trying to create a new image from a container, but with a different `RUN` command. That may be a perversion of using containers to build images. The [road map for Builder](https://github.com/docker/docker/issues/735#issuecomment-120493444) seems promising (and cleaner).

**Comment by cpuguy83** [comment]
@posita This is what `CMD` and `ENTRYPOINT` are for. `RUN` is the command that was used to create the container that you are commiting.

**Comment by posita** [comment]
@cpuguy83, sorry, I didn't articulate very well. I'm thinking more along the lines of dynamic image inheritance, which a builder API as planned appears to accommodate without making a mess of things. My apologies for the confusion. In any event, I see now that this issue is properly resolved by #14301 as far as I'm concerned. I will wait for the `cp` command and the new builder API.
